Test updates (#505)
Azure/azure-iot-cli-extension: CONTRIBUTING.md

`azext_dt_ep_servicebus_topic` is the topic in the Service Bus Namespace to use. Optional variable, used for Digital Twin endpoint tests.

`azext_dt_ep_eventhub_namespace` is the Event Hub Namespace to use. Optional variable, used for Digital Twin endpoint and data history tests.

`azext_dt_ep_eventhub_policy` is the policy for the Event Hub instance in the Event Hub Namespace to use. Optional variable, used for Digital Twin endpoint and data history tests.

`azext_dt_ep_eventhub_topic` is the Event Hub instance in the Event Hub Namespace to use. Optional variable, used for Digital Twin endpoint and data history tests.

`azext_dt_ep_eventhub_topic_consumer_group` is the Service Bus Namespace to use. Optional variable, used for Digital Twin data history tests.

`azext_dt_ep_rg` is the resource group that contains the endpoint (Event Hub, Event Grid, Service Bus) variables. Optional variable, if not provided, `az_iot_testrg` will be used as the resource group.

##### IoT Hub

Execute the following command to run the IoT Hub integration tests:

> `pytest azext_iot/tests/iothub/ -k "_int.py"`

Reviewer: The `.py` suffix shouldn't be strictly necessary to run the IT's.

Author: without the ".py", some unit tests get caught because they have "_int" in the test name (although this only happens for central tests, I got used to adding the py):
![image](https://user-images.githubusercontent.com/73560279/161867657-38112a15-8b74-48a0-a22f-c6773162437f.png)

If we say that the keyword "_int" must only match unit tests, it will restrict naming unit tests so I prefer keeping the extension.

Reviewer: Ok - the line I commented on has a path constraint for iothub, which doesn't have this issue (like you pointed out its currently an issue with central tests).

Then why include the extension path _only_ for a service-test area that doesn't have the issue? Further down there is still usage of `_int`

![image](https://user-images.githubusercontent.com/28658112/161871821-6c973789-d159-44a5-8edb-7d5fc4f58372.png)

pip install -r path/to/source/dev_requirements
```

Example unit tests runs:

_Hub:_
`pytest azext_iot/tests/iothub/core/test_iot_ext_unit.py`

_DPS:_
`pytest azext_iot/tests/dps/core/test_dps_discovery_unit.py`

Unit tests end in "_unit" so execute the following command to run all unit tests,

> `pytest -k "_unit"`

Reviewer: Should the unit test examples not have the file extension based on the concern raised on the integration test examples? I think they should be consistent.

Execute the following command to run the IoT Hub unit tests:

`pytest azext_iot/tests/iothub/ -k "_unit"`

### Integration Tests

Integration tests are run against Azure resources and depend on environment variables.

#### Azure Resource Setup

The following resources will be needed for the integration tests.

- IoT Hub
- IoT Device Provisioning Service

> - IoT Digital Twin

Reviewer: ```suggestion
- Azure Digital Twins instance
```
The official marketing for the Azure Digital Twins service usually doesn't have an IoT prefix https://docs.microsoft.com/en-us/azure/digital-twins/overview

- IoT Central App
- Storage Account (with an empty Container)
- Event Grid Topic
- Event Hub Namespace with an Event Hub
- Service Bus Namespace with a Topic
- Azure Data Explorer Cluster with a Database

If specified in the pytest.ini configuration file, those resources will be used. Please ensure that the resources are in a clean, new state (ex: Iot Hub should not have any devices). Otherwise, new resources will be generated during the test startup and deleted during the test teardown.

> Note that if you stop the code with ctrl + C, the resources may not be deleted properly.

Reviewer: ```suggestion
> Note: If you interrupt test execution (for example via ctrl + C), the teardown or clean up processes may not run leaving resources in an indeterminant state.
```

    azext_iot_central_storage_cstring=
    azext_iot_central_storage_container=
    azext_dt_adx_cluster=
    azext_dt_adx_database=
    azext_dt_adx_rg=
    azext_dt_ep_eventgrid_topic=
    azext_dt_ep_servicebus_namespace=
    azext_dt_ep_servicebus_policy=
    azext_dt_ep_servicebus_topic=
    azext_dt_ep_eventhub_namespace=
    azext_dt_ep_eventhub_policy=
    azext_dt_ep_eventhub_topic=
    azext_dt_ep_eventhub_topic_consumer_group=
    azext_dt_ep_rg=
    azext_dt_region=
```

To run almost all of the tests, only the `azext_iot_testrg` is needed.

To run all tests, `azext_iot_testrg`, `azext_iot_central_app_id`, and `azext_iot_central_primarykey` are needed because the IoT Central Primary Key cannot be currently retrieved through the CLI.

For all resources, if the environmental variable is not provided, a new instance will be created for the test run and deleted at the end of the test run.

##### General Variables

`az_iot_testrg` is the resource group that contains the IoT Hub and DPS instances or where all test resources are created. This is required, as it will be the default resource group if any other resource group variables are not provided.

> `azext_iot_testhub` is the test IoT Hub name. Optional variable, used for IoT Hub and DPS tests. If not provided, a new IoT Hub will be created for the test run (and deleted at the end of the test).

Reviewer: `azext_iot_testhub` should probably be in an 'IoT Hub test variables' category

`azext_iot_testdps` should probably be in an 'IoT Device Provisioning Service' category

`azext_iot_testdps` is the test DPS name. Optional variable used for IoT DPS tests. If not provided, a new DPS instance will be created for the test run (and deleted at the end of the test).

`azext_iot_teststorageaccount` is the storage account used for running IoT Hub and Central storage tests. Optional variable, specify only when you want to run storage tests. During these tests, your hub will be assigned a System-Assigned AAD identity, and will be granted the role of "Storage Blob Data Contributor" on the storage account you provide. Both the hub's identity and the RBAC role will be removed once the test completes. No role assignments are made for the IoT Central App.

`azext_iot_teststoragecontainer` is the name of blob container belonging to the above mentioned storage account. Optional environment variable, defaults to 'devices' when not specified.

> ##### IoT Central Test variables

Reviewer: What about a markdown table to organize the vars for a sharper presentation? Columns could be  corresponding to the var  name, required and description.

`azext_dt_ep_eventhub_namespace` is the Event Hub Namespace to use. Optional variable, used for Digital Twin endpoint and data history tests.

`azext_dt_ep_eventhub_policy` is the policy for the Event Hub instance in the Event Hub Namespace to use. Optional variable, used for Digital Twin endpoint and data history tests.

`azext_dt_ep_eventhub_topic` is the Event Hub instance in the Event Hub Namespace to use. Optional variable, used for Digital Twin endpoint and data history tests.

`azext_dt_ep_eventhub_topic_consumer_group` is the Service Bus Namespace to use. Optional variable, used for Digital Twin data history tests.

`azext_dt_ep_rg` is the resource group that contains the endpoint (Event Hub, Event Grid, Service Bus) variables. Optional variable, if not provided, `az_iot_testrg` will be used as the resource group.

##### IoT Hub

Execute the following command to run the IoT Hub integration tests:

`pytest azext_iot/tests/iothub/ -k "_int.py"`

##### Device Provisioning Service

Execute the following command to run the IoT Hub DPS integration tests:

> `pytest azext_iot/tests/dps/ -k "_int.py"`

Reviewer: Thoughts about combining the 'executing the service tests' sections with the 'setting the service specific env vars to run tests' sections?

Author: I reformatted how the tests are presented. Please look over the changes

pip install -r path/to/source/dev_requirements
```

Example unit tests runs:

_Hub:_
`pytest azext_iot/tests/iothub/core/test_iot_ext_unit.py`

_DPS:_
`pytest azext_iot/tests/dps/core/test_dps_discovery_unit.py`

Unit tests end in "_unit.py" so execute the following command to run all unit tests,
`pytest -k "_unit.py"`

Execute the following command to run the IoT Hub unit tests:

`pytest azext_iot/tests/iothub/ -k "_unit.py"`

### Integration Tests

Integration tests are run against Azure resources and depend on environment variables.

Example int tests runs:

_Hub:_
`pytest azext_iot/tests/iothub/core/test_iothub_storage_int.py`

_DPS:_
`pytest azext_iot/tests/dps/core/test_dps_discovery_int.py`

Integration tests end in "_int_.py" so execute the following command to run all integration tests,

> `pytest -k "_int_.py"`

Reviewer: \_int_.py?

| `azext_dt_ep_eventhub_namespace`	| Digital Twin Data History and Endpoint Tests	| The Event Hub Namespace to use.	|
| `azext_dt_ep_eventhub_policy`	| Digital Twin Endpoint Tests	| The policy for the Event Hub instance in the Event Hub Namespace to use.	|
| `azext_dt_ep_eventhub_topic`	| Digital Twin Data History and Endpoint Tests	| The Event Hub instance in the Event Hub Namespace to use.	|
| `azext_dt_ep_eventhub_topic_consumer_group`	| Digital Twin Data History Tests	| The Event Hub consumer group name to use. Defaults to "$Default".	|
| `azext_dt_ep_rg`	| Digital Twin Data History and Endpoint Tests	| The resource group that contains the endpoint (Event Hub, Event Grid, Service Bus) variables. If not provided, `az_iot_testrg` will be used as the resource group.	|

##### IoT Central

IoT Central integration tests can be run against a specific available api versions using command line argument "--api-version"

e.g. run tests against v 1.0

`pytest azext_iot/tests/central/ -k "_int.py" --api-version "1.0"`

If the "--api-version" argument is not specified, all runs act against default api version for each tested command.

##### Test Resource Tagging

> There are 4 more test variables used for tagging the test resources:

Reviewer: Not including these test variables in the IT variables table or list above?

Author: I am trying to show that these are separate and would be used by internal developers (as in us)

Author: updated

| `azext_dt_ep_eventhub_topic_consumer_group`	| Digital Twin Data History Tests	| The Event Hub consumer group name to use. Defaults to "$Default".	|
| `azext_dt_ep_rg`	| Digital Twin Data History and Endpoint Tests	| The resource group that contains the endpoint (Event Hub, Event Grid, Service Bus) variables. If not provided, `az_iot_testrg` will be used as the resource group.	|
| `use_tags`	| IoT Hub, DPS, and Central Tagging	| Flag to enable resource tagging. Please see "Test Resource Tagging" for more details.	|
| `definition_id`	| IoT Hub, DPS, and Central Tagging	| Definition Id populated by an internal pipeline run. Can be manually set to customize the pipeline name tag. Please see "Test Resource Tagging" for more details.	|
| `job_display_name`	| IoT Hub, DPS, and Central Tagging	| Job Display Name populated by an internal pipeline run. Can be manually set to customize the pipeline name tag. Please see "Test Resource Tagging" for more details.	|
| `job_id`	| IoT Hub, DPS, and Central Tagging	| Job Id populated by an internal pipeline run. Can be manually set to customize the pipeline name tag. Please see "Test Resource Tagging" for more details.	|

##### IoT Central

> IoT Central integration tests can be run against a specific available api versions using command line argument "--api-version"

Reviewer: ```suggestion
IoT Central integration tests can be run against a specific available api version using command line argument "--api-version"
```

> IoT Central integration tests can be run against a specific available api versions using command line argument "--api-version"

Reviewer: ```suggestion
IoT Central integration tests can be run against a specific available api version using command line argument "--api-version"
```
